pulsar-common/src/test/java/org/apache/pulsar/common/nar/NarUnpackerTest.java

```diff
@@ -0,0 +1,160 @@
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+@Slf4j
+public class NarUnpackerTest {
+    File sampleZipFile;
+    File extractDirectory;
+
+    @BeforeMethod
+    public void createSampleZipFile() throws IOException {
+        sampleZipFile = Files.createTempFile("sample", ".zip").toFile();
+        try (ZipOutputStream out = new ZipOutputStream(new FileOutputStream(sampleZipFile))) {
+            for (int i = 0; i < 10000; i++) {
+                ZipEntry e = new ZipEntry("hello" + i + ".txt");
+                out.putNextEntry(e);
+                byte[] msg = "hello world!".getBytes(StandardCharsets.UTF_8);
+                out.write(msg, 0, msg.length);
+                out.closeEntry();
+            }
+        }
+        extractDirectory = Files.createTempDirectory("nar_unpack_dir").toFile();
+    }
+
+    @AfterMethod(alwaysRun = true)
+    void deleteSampleZipFile() throws IOException {
+        if (sampleZipFile != null) {
+            sampleZipFile.delete();
+        }
+        if (extractDirectory != null) {
+            FileUtils.deleteFile(extractDirectory, true);
+        }
+    }
+
+    @Test
+    void shouldExtractFilesOnceInSameProcess() throws InterruptedException {
+        int threads = 20;
+        CountDownLatch countDownLatch = new CountDownLatch(threads);
+        AtomicInteger exceptionCounter = new AtomicInteger();
+        AtomicInteger extractCounter = new AtomicInteger();
+        for (int i = 0; i < threads; i++) {
+            new Thread(() -> {
+                try {
+                    NarUnpacker.doUnpackNar(sampleZipFile, extractDirectory, extractCounter::incrementAndGet);
+                } catch (Exception e) {
+                    e.printStackTrace();
+                    exceptionCounter.incrementAndGet();
+                } finally {
+                    countDownLatch.countDown();
+                }
+            }).start();
+        }
+        assertTrue(countDownLatch.await(30, TimeUnit.SECONDS));
+        assertEquals(exceptionCounter.get(), 0);
+        assertEquals(extractCounter.get(), 1);
+    }
+
+    public static class NarUnpackerWorker {
+        public static void main(String[] args) {
+            File sampleZipFile = new File(args[0]);
+            File extractDirectory = new File(args[1]);
+            AtomicInteger extractCounter = new AtomicInteger();
+            try {
+                NarUnpacker.doUnpackNar(sampleZipFile, extractDirectory, extractCounter::incrementAndGet);
+                if (extractCounter.get() == 1) {
+                    System.exit(101);
+                } else if (extractCounter.get() == 0) {
+                    System.exit(100);
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+                System.exit(99);
+            }
+        }
+    }
+
+    @Test
+    void shouldExtractFilesOnceInDifferentProcess() throws InterruptedException {
```

Comment:
I am not sure it is worth to add such test.
it is quite heavyweight (especially for CI, but also for people that are running the tests for instance during a release validation) and we are not sure it reproduces consistently the problem.

I suggest to drop this test but if you feel strong to keep it then keep it

Reply:
it's necessary for validating the behavior of cross-process locks. It's very lightweight compared to many other tests.